### PR TITLE
Introduce Wallet Select Dropdown

### PR DESCRIPTION
**Pull request by johnny9**
WalletSelect is a Popup that appears after clicking the main
WalletBadge in the DesktopNavigation bar. It contains a ListView
that allows the user to select one of the wallets listed in the
wallet directory.

This PR uses the arrow icon that is shared with the Tooltip so it is based off of that commit.

[Build Artifacts](https://github.com/bitcoin-core/gui-qml/actions/runs/9263281845)

### CODE REVIEW

File: src/qml/models/walletlistmodel.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#include <qml/models/walletlistmodel.h>
```
Comment (MarnixCroes) [inline src/qml/models/walletlistmodel.cpp:1]: forgot to add copyright here?

**Review by MarnixCroes — COMMENTED** [review]
two nits
- wallet list should be closed when clicking outside of it. i.e. navigate to settings -> wallet list is still shown
- small alignment 
image

**Comment by johnny9** [comment]
Update  from 4ef3149 to 5d191d1 
 - Updated include guard on wallet list model
 - Added copyright header to walletlistmodel.cpp
 - Adjusted padding on Add Wallet icon
 - Changed closePolicy on WalletSelect Popup to CloseOnPressOutside
 - Rebased with main

**Comment by GBKS** [comment]
> so what's that "X" for? 

The idea is that you can close the wallet that is currently open (result is that there would be no open wallet). And if you press "x" on a close wallet, you would remove it from the application altogether. We don't have those anywhere else in the UI yet on mobile.

Are there other options you may want to directly take on the wallet from here? Like changing the name? If we have more, we could instead use an ellipsis `...` to access a sub-menu.

**Review by D33r-Gee — COMMENTED** [review]
tACK on WSL Ubuntu 22.04

Looks like it behaves the way it's intended (see screenshot below)

<details>
<summary>screenshot</summary>

Screenshot 2024-06-07 145925

</details>

**Comment by hebasto** [comment]
@pablomartin4btc 
> tACK [4ef3149](https://github.com/bitcoin-core/gui-qml/commit/4ef3149cc035a9f7b52fb7ebd90ab1a902d83795)

Want to re-ACK?

**Comment by hebasto** [comment]
@D33r-Gee 
> tACK on WSL Ubuntu 22.04

To ensure that someone's ACK is related to a certain version of the branch, it is wise to mention the top commit hash. Besides, the merge script needs a top commit hash to count an ACK.

**Comment by pablomartin4btc** [comment]
> > so what's that "X" for?
> 
> The idea is that you can close the wallet that is currently open (result is that there would be no open wallet). And if you press "x" on a close wallet, you would remove it from the application altogether. We don't have those anywhere else in the UI yet on mobile.

Mmm... then we need to indicate somehow when the wallet is open and the "x" would close it and when it's closed and the "x" will remove it from the UI... and how do you get it back when has been removed? It's not very intuitive at the moment.

Also, I'm not sure if listing all wallets available in the datadir would be practical... perhaps this can be configured as an option and disabled as default or if there are more then 5/ 10 wallets asking the user and persist that into the config?

> Are there other options you may want to directly take on the wallet from here? Like changing the name? If we have more, we could instead use an ellipsis `...` to access a sub-menu.

Yeah, there's no "renaming" feature in `QT` currently, it could be useful and we can put the "remove" from the app or "close" the wallet... still, how do you added back once removed?

**Comment by pablomartin4btc** [comment]
> @pablomartin4btc
> 
> > tACK [4ef3149](https://github.com/bitcoin-core/gui-qml/commit/4ef3149cc035a9f7b52fb7ebd90ab1a902d83795)
> 
> Want to re-ACK?

Checking... somehow I missed 5d191d1bedda3f10ab4457518cd5a5f249a23d73.

**Review by pablomartin4btc — COMMENTED** [review]
tACK 5d191d1bedda3f10ab4457518cd5a5f249a23d73

I still have these [concerns](https://github.com/bitcoin-core/gui-qml/pull/401#issuecomment-2156452481).

**Comment by GBKS** [comment]
Pablo and I spoke about the wallet selector and I did another design iteration based on my current understanding. Here's a [video](https://www.youtube.com/watch?v=LCv6B0rgnc8) that covers it. Still have some minor questions in there, please take a look and let me know how it looks.

**Review by pablomartin4btc — COMMENTED** [review]
> Here's a [video](https://www.youtube.com/watch?v=LCv6B0rgnc8) that covers it.

These are my obs:

- ordering the wallet list on last opened criteria (last 5? then alphabetical?);

- scroll solution ok (but we need to solve the bar that gets on top of the remove/ close wallet icons);

- icon for remove wallet looks fine but for closing looks a bit simple (what about using the wallet icon at the top when there are no wallets at all?)... also, what if we put both icons? you dont need to close (unload) the wallet first and then remove it, no?

- once we remove a wallet, perhaps after the user clicks on the "Add +" we show these options: create a new or "add previously removed walet" or "link a wallet dir"?

- we still have the issue when there are many wallets, what about a warning? "you have more than 15 wallets in your datadir, do you want to list them all? (and something saying this will be asked once and saved the configuration into the settings?);

These are open questions to the other reviewers as well.

**Comment by GBKS** [comment]
Thanks for the feedback.

> ordering the wallet list on last opened criteria (last 5? then alphabetical?);

Why not always use the last open date?

> scroll solution

I will mock this up

> icon for remove wallet looks fine but for closing looks a bit simple...

I'll do another revision of this design as we discussed yesterday, where the wallet options are in a small overlay menu

> once we remove a wallet, perhaps after the user clicks on the "Add +" we show these options: create a new or "add previously removed walet" or "link a wallet dir"?

The current design is that in the "Add wallet" screen, you can import. From there you select the wallet file you want to add. Does that work?

> we still have the issue when there are many wallets

Why is this an issue?

**Comment by pablomartin4btc** [comment]
> > ordering the wallet list on last opened criteria (last 5? then alphabetical?);
> 
> Why not always use the last open date?

Depends on the amount of wallets we are talking about, that's what I call "issue", perhaps on the mobile version doesn't make sense to have as many, not sure. I'll ask other devs.

> > once we remove a wallet, perhaps after the user clicks on the "Add +" we show these options: create a new or "add previously removed walet" or "link a wallet dir"?
> 
> The current design is that in the "Add wallet" screen, you can import. From there you select the wallet file you want to add. Does that work?

Ok, maybe that's the use case. Is it possible for you to do a prototype as you did in the previous video to see how it would work? From removing the wallet from the list to add it back (if it's doable with the tool you are using).
 
> > we still have the issue when there are many wallets
> 
> Why is this an issue?

Performance will be impacted loading multiple wallets but what I'm not convinced about is building the list of wallets in the menu with whatever amount of wallets there are in the data directory. I'll ask around, perhaps there's nothing wrong with it and I'm just overly worried.

**Comment by D33r-Gee** [comment]
re-tACK [5d191d1](https://github.com/bitcoin-core/gui-qml/commit/5d191d1bedda3f10ab4457518cd5a5f249a23d73) on WSL Ubuntu 22.04 works as expected...

Does it need rebase?

File: src/qml/pages/wallet/DesktopWallets.qml [hunk 2]
```diff
@@ -21,27 +21,26 @@ Page {
...
+                onClicked: {
+                    walletListModel.listWalletDir()
+                    walletSelect.opened ? walletSelect.close() : walletSelect.open()
                 }
             }
+
+            WalletSelect {
+                id: walletSelect
+                model: walletListModel
+                closePolicy: Popup.CloseOnPressOutside
```
Comment (MarnixCroes) [inline src/qml/pages/wallet/DesktopWallets.qml:40]: close when press outside.
I guess it should also close when clicking on it again, additionally when clicking outside?

**Review by MarnixCroes — COMMENTED** [review]
5d191d1bedda3f10ab4457518cd5a5f249a23d73
some possible improvements (maybe for follow up):
- when having many wallets there should be some scrollbar, also to be able to select wallets displayed at the bottom
image

---

- the wallet name should not break out of the width of the wallet select pop up
image

File: src/qml/models/walletlistmodel.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    roles[NameRole] = "name";
+    return roles;
+}
+
+void WalletListModel::addItem(const Item &item)
+{
+    beginInsertRows(QModelIndex(), rowCount(), rowCount());
+    m_items.append(item);
+    endInsertRows();
+}
```
Comment (D33r-Gee) [inline src/qml/models/walletlistmodel.cpp:81]: Does it need a blank line at the end?

**Comment by D33r-Gee** [comment]
tACK https://github.com/bitcoin-core/gui-qml/commit/411a08a55bfeba795ede692233134e31ca58cbe7 on WSL Ubuntu 22.04

Works as expected...

<details>
  <summary>Ubuntu 22.04 Screenshot</summary>
  
Screenshot 2024-07-25 135609

</details>

**Review by MarnixCroes — APPROVED** [review]
tACK 833fc38e56da2914877c672280d3606c3d2bb0ba
previous points https://github.com/bitcoin-core/gui-qml/pull/401#pullrequestreview-2158382120 can be done as follow up

File: src/qml/pages/wallet/WalletBadge.qml [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            PropertyChanges { target: icon; color: root.textActiveColor }
+        },
+        State {
+            name: "HOVER"; when: root.hovered
+            PropertyChanges { target: buttonText; color: root.textHoverColor }
+            PropertyChanges { target: icon; color: root.textHoverColor }
+            PropertyChanges { target: balanceText; color: root.textHoverColor }
+        }
+    ]
+}
```
Comment (D33r-Gee) [inline src/qml/pages/wallet/WalletBadge.qml:151]: does it need a blank line at the bottom?

File: src/qml/pages/wallet/DesktopWallets.qml (hunk 2 above)
Comment (johnny9) [reply src/qml/pages/wallet/DesktopWallets.qml:40]: I want to add this as an issue. It turns out its more difficult than I thought to fix this.

File: src/qml/pages/wallet/WalletBadge.qml [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                    return '<font color="' + highlightColor + '">' + match + '</font>';
+                }
+                return match;
+            });
+        }
+
+        // Concatenate the parts back together
+        var formattedBitcoins = parts.join('.');
+
+        // Format the text with the Bitcoin symbo
```
Comment (MarnixCroes) [inline src/qml/pages/wallet/WalletBadge.qml:69]: nit
```suggestion
        // Format the text with the Bitcoin symbol
```

**Review by MarnixCroes — APPROVED** [review]
tACK 366c605b585c3519fe8e6ad39374d283b91ac071

some nits, can be follow up: 
- a visible scrollbar for the wallet list
- when a wallet name exceeds the width and is thus cut off -> would be good display the full wallet name in a tooltip on hover

**Comment by johnny9** [comment]
Update  from 366c605 to 0939f2b 
- Fixed typo in comment in WalletBadge.qml

**Comment by D33r-Gee** [comment]
tACK [0939f2b](https://github.com/bitcoin-core/gui-qml/commit/0939f2bfd7586c04ad35aabffc5028511c1a236f) on WSL Ubuntu 22.04

Works as expected and looks great

<details>
  <summary>Ubuntu Screenshot</summary>

Screenshot 2024-08-09 123135


</details>

**Review by MarnixCroes — APPROVED** [review]
re-ack 0939f2bfd7586c04ad35aabffc5028511c1a236f

**Comment by hebasto** [comment]
> re-ack [0939f2b](https://github.com/bitcoin-core/gui-qml/commit/0939f2bfd7586c04ad35aabffc5028511c1a236f)

FWIW, "ack" is not counted by the merge script. The "ACK" is expected.

**Pull request merged**